src/Typecheck.elm

```diff
@@ -107,31 +108,75 @@ checkSig sig args =
       Nothing -> Invalid
 
 
+substitute : Term -> Term -> String -> Term
+substitute old new n =
+  case old of
+    CTerm c -> CTerm c
+    VTerm v -> if n == v then new else VTerm v
+    Lam t1 t2 ->
+      case t1 of
+        VTerm v -> if n == v then Lam t1 t2 else Lam t1 (substitute t2 new n)
+        _ -> Lam t1 t2
+    App t1 t2 -> App (substitute t1 new n) (substitute t2 new n)
+    Plus t1 t2 -> Plus (substitute t1 new n) (substitute t2 new n)
+    Minus t1 t2 -> Minus (substitute t1 new n) (substitute t2 new n)
+    Times t1 t2 -> Times (substitute t1 new n) (substitute t2 new n)
+    Eq t1 t2 -> Eq (substitute t1 new n) (substitute t2 new n)
+    And t1 t2 -> And (substitute t1 new n) (substitute t2 new n)
+    Or t1 t2 -> Or (substitute t1 new n) (substitute t2 new n)
+    any -> any
+
+
+typeCheckApp : Env -> Term -> CheckResult
+typeCheckApp env t =
+  case t of
+    App x y ->
+      case x of
+        Lam w z ->
+          case w of
+            VTerm n -> typecheck env (substitute z y n)
+            _ -> Invalid
+        VTerm v ->
+          let lambda = lookup env v
+            in case lambda of
+              Just (Lam w z) ->
+                case w of
+                  VTerm n -> typecheck env (substitute z y n)
+                  _ -> Invalid
+              _ -> Invalid
+        _ -> Invalid
+    _ -> Invalid
+
+
+getTypeSignature : Env -> Term -> List VType
+getTypeSignature env t =
+  case t of
+    CTerm c ->
+      case c of
+        CBool _ -> [TBool]
+        CInt _  -> [TInt]
+
+    VTerm v ->
+      case lookup env v of
+        Just sub -> []
+        Nothing -> []
+
```

Comment:
I originally wrote this rendition of type signatures with the idea that we would be representing types as a list of basic types.  Since our discussion Wednesday however, I think we're moving to a better understanding of function types as being built up from our `TLam` type (which by the way looks like a good step in the right direction).

As a consequence, our typecheck may need a major rewrite (eventually).  For instance, with this new scheme, I believe the proper type of `Plus` should be `TLam VInt (TLam VInt VInt)`, which would allow us to define this operator within the constructs of the langauge.

For now I think these changes are good to be merged in, but I think this refactoring should happen soon.